# Walmart Sales Data Analysis Project With MySQL

# Author: Benard Mwinzi

## Summary

This project aims to use MySQL to analyze Walmart's sales data and extract meaningful insights to solve business problems. The Walmart sales data contains 17 columns like InvoiceID, Branch, City, Customer_type, Gender, Product_line, Unit_price, Quantity, Tax_5_Percent, Total, Date, Time, Payment, cogs, gross_margin_percentage, gross_income and Rating. The data was obtained from a kaggle.com competition: https://www.kaggle.com/c/walmart-recruiting-store-sales-forecasting

The analysis was conducted in three sections: product, sales, and customer analysis.

To achieve the objectives, the data must first be cleaned by checking for missing values, creating new columns, and changing data types if necessary.

Finally, an exploratory data analysis will be conducted to answer research questions.

### Products

i) How many unique product lines does the data have?

ii) What is the most common payment method?

iii) What is the most selling product line?

iv) What is the total revenue by month?

v) What month had the largest COGS?


vi)	What product line had the largest revenue?

vii) What is the city with the largest revenue?

viii) What product line had the largest VAT?

ix) Fetch each product line and add a column to those product line showing "Good", "Bad". Good if its greater than average sales.

x) Which branch sold more products than average product sold?

xi) What is the most common product line by gender?

xii) What is the average rating of each product line? 


#### *Sales*

i) What is the number of sales made in each time of the day per weekday?

ii) Which of the customer types brings the most revenue?

iii) Which city has the largest tax percentage/ VAT (Value Added Tax)?

iv) Which customer type pays the most VAT?

#### *Customers*

i) How many unique customer types does the data have?

ii) How many unique payment methods does the data have?

iii) What is the most common customer type?

iv) What is the gender of most of the customers?

v) What is the gender distribution per branch?

vi) Which time of the day do customers give the most ratings?

vii) At what time of the day do customers give the most ratings per branch?

viii) Which day of the week has the best average ratings?

ix) Which day of the week has the best average ratings per branch?

# Connecting to the Database and Loading the Data

In [1]:
#Loading necessary libraries
import pymysql
import pandas as pd

In [3]:
#Connecting to my Database
# % -- 
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [4]:
#connecting to the database
%sql mysql+mysqldb://root:YourPassword@localhost/Walmart

### Creating an sql table to load the data

In [4]:
%%sql
DROP TABLE IF EXISTS Walmart;

 * mysql+mysqldb://root:***@localhost/Walmart
0 rows affected.


[]

In [5]:
%%sql
CREATE TABLE Walmart (
    InvoiceID VARCHAR (20) NOT NULL PRIMARY KEY,
    Branch VARCHAR (2) NOT NULL,
    City VARCHAR (30) NOT NULL,
    Customer_type VARCHAR (20) NOT NULL,
    Gender VARCHAR (10) NOT NULL,
    Product_line VARCHAR (30) NOT NULL,
    Unit_price DECIMAL (7,2) NOT NULL,
    Quantity INT,
    Tax_5_Percent DECIMAL (6,4) NOT NULL,
    Total DECIMAL (15,4) NOT NULL,
    Date DATE NOT NULL,
    Time TIME NOT NULL,
    Payment VARCHAR (15) NOT NULL,
    cogs DECIMAL (10,2) NOT NULL,
    gross_margin_percentage DECIMAL (20,10),
    gross_income DECIMAL (12,4),
    Rating DECIMAL (5,2));

 * mysql+mysqldb://root:***@localhost/Walmart
0 rows affected.


[]

### Importing the data into the table

In [6]:
#Removing the strict sql mode for the session before importing the data to enable mysql to import black spaces in decimal data
%sql SET SESSION sql_mode = ''

 * mysql+mysqldb://root:***@localhost/Walmart
0 rows affected.


[]

Loading the data into the table

In [7]:
%%sql
LOAD DATA INFILE 'D:/Data Science/MySQL/Walmart Project/WalmartSalesData.csv'
INTO TABLE Walmart
FIELDS TERMINATED BY ',' ENCLOSED BY '"'
LINES TERMINATED BY '\n'
IGNORE 1 LINES;

 * mysql+mysqldb://root:***@localhost/Walmart
1000 rows affected.


[]

Our dataset has 1,000 rows

## Data Cleaning
The data was cleaned by ensuring that there was no missing values.

To do that, we first needed to see the first five rows of the dataset.

In [8]:
%%sql
SELECT * 
FROM walmart
LIMIT 5;

 * mysql+mysqldb://root:***@localhost/Walmart
5 rows affected.


[('101-17-6199', 'A', 'Yangon', 'Normal', 'Male', 'Food and beverages', Decimal('45.79'), 7, Decimal('16.0265'), Decimal('336.5565'), datetime.date(2019, 3, 13), datetime.timedelta(seconds=71040), 'Credit card', Decimal('320.53'), Decimal('4.7619047620'), Decimal('16.0265'), Decimal('7.00')),
 ('101-81-4070', 'C', 'Naypyitaw', 'Member', 'Female', 'Health and beauty', Decimal('62.82'), 2, Decimal('6.2820'), Decimal('131.9220'), datetime.date(2019, 1, 17), datetime.timedelta(seconds=45360), 'Ewallet', Decimal('125.64'), Decimal('4.7619047620'), Decimal('6.2820'), Decimal('4.90')),
 ('102-06-2002', 'C', 'Naypyitaw', 'Member', 'Male', 'Sports and travel', Decimal('25.25'), 5, Decimal('6.3125'), Decimal('132.5625'), datetime.date(2019, 3, 20), datetime.timedelta(seconds=64320), 'Cash', Decimal('126.25'), Decimal('4.7619047620'), Decimal('6.3125'), Decimal('6.10')),
 ('102-77-2261', 'C', 'Naypyitaw', 'Member', 'Male', 'Health and beauty', Decimal('65.31'), 7, Decimal('22.8585'), Decimal('480.0285'), datetime.date(2019, 3, 5), datetime.timedelta(seconds=64920), 'Credit card', Decimal('457.17'), Decimal('4.7619047620'), Decimal('22.8585'), Decimal('4.20')),
 ('105-10-6182', 'A', 'Yangon', 'Member', 'Male', 'Fashion accessories', Decimal('21.48'), 2, Decimal('2.1480'), Decimal('45.1080'), datetime.date(2019, 2, 27), datetime.timedelta(seconds=44520), 'Ewallet', Decimal('42.96'), Decimal('4.7619047620'), Decimal('2.1480'), Decimal('6.60'))]

#### *Dealing with Missing Values*

We need to check if the data has null values. Since we already used the NOT NULL on some variables when creating the table, we are sure that those variables do not contain any missing values.

We must check for missing vales in gross_margin_percentage, gross_income, and Rating variables.


In [9]:
%%sql
SELECT * 
FROM Walmart
WHERE gross_margin_percentage = "" OR gross_income = "" OR Rating = 0;

 * mysql+mysqldb://root:***@localhost/Walmart
0 rows affected.


[]

The above query seeks to get any missing values from any of the three variables. The query returned 0 rows meaning that there is no row with missing values.

#### *Creating New Variables*

We need to add a Month colum by extracting month name from the Date column.

To do that, we first alter the table to create a new month column

In [10]:
%%sql
ALTER TABLE Walmart
ADD Month VARCHAR (15);

 * mysql+mysqldb://root:***@localhost/Walmart
0 rows affected.


[]

Now, we update the table to populate the Month column by extracting month name from the Date column

In [11]:
%%sql
UPDATE Walmart
SET Month = MONTHNAME(Date);

 * mysql+mysqldb://root:***@localhost/Walmart
1000 rows affected.


[]

Checking that our column has been populated as neeeded.

In [12]:
%%sql
SELECT *
FROM Walmart
LIMIT 3;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('101-17-6199', 'A', 'Yangon', 'Normal', 'Male', 'Food and beverages', Decimal('45.79'), 7, Decimal('16.0265'), Decimal('336.5565'), datetime.date(2019, 3, 13), datetime.timedelta(seconds=71040), 'Credit card', Decimal('320.53'), Decimal('4.7619047620'), Decimal('16.0265'), Decimal('7.00'), 'March'),
 ('101-81-4070', 'C', 'Naypyitaw', 'Member', 'Female', 'Health and beauty', Decimal('62.82'), 2, Decimal('6.2820'), Decimal('131.9220'), datetime.date(2019, 1, 17), datetime.timedelta(seconds=45360), 'Ewallet', Decimal('125.64'), Decimal('4.7619047620'), Decimal('6.2820'), Decimal('4.90'), 'January'),
 ('102-06-2002', 'C', 'Naypyitaw', 'Member', 'Male', 'Sports and travel', Decimal('25.25'), 5, Decimal('6.3125'), Decimal('132.5625'), datetime.date(2019, 3, 20), datetime.timedelta(seconds=64320), 'Cash', Decimal('126.25'), Decimal('4.7619047620'), Decimal('6.3125'), Decimal('6.10'), 'March')]

As seen in the above output, the month column has been appropriately created.

# Exploratory Data Analysis

The exploratory data analysis was conducted to answer the research questions and understand patterns and trends in company sales.

The EDA was conducted to understand product, Sales, and Customers. 

## Product 

This part of the EDA aims at understanding the trends and patterns in the company's products.

#### *i)	How many unique product lines does the data have?*

In [13]:
%%sql
SELECT DISTINCT Product_line 
FROM Walmart;

 * mysql+mysqldb://root:***@localhost/Walmart
6 rows affected.


[('Food and beverages',),
 ('Health and beauty',),
 ('Sports and travel',),
 ('Fashion accessories',),
 ('Home and lifestyle',),
 ('Electronic accessories',)]

The query returned 6 rows, which means that there are 6 unique product lines. 

Alternatively, we count the number of distinct product lines instead of checking the number of rows retuned.

In [14]:
%%sql
SELECT COUNT(DISTINCT Product_line) AS UniqueProductLines
FROM Walmart;

 * mysql+mysqldb://root:***@localhost/Walmart
1 rows affected.


[(6,)]

There are six unique product lines

#### *ii)	What is the most common payment method?*

Here, we need to select the count of invoiceIDs and group them based on the payment methods.

In [34]:
%%sql
SELECT Payment, COUNT(InvoiceID) AS NoOfPayments
FROM Walmart
GROUP BY Payment
ORDER BY NoOfPayments DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('Ewallet', 345), ('Cash', 344), ('Credit card', 311)]

EWallet is the most common payment method. It was used to pay 345 invoices.

#### *iii) What is the most selling product line?*

Here, we select the sum of quantity grouped by the product line.

In [16]:
%%sql
SELECT Product_line, SUM(Quantity) AS TotalQuantity
FROM Walmart
GROUP BY Product_line
ORDER BY TotalQuantity DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
6 rows affected.


[('Electronic accessories', Decimal('971')),
 ('Food and beverages', Decimal('952')),
 ('Sports and travel', Decimal('920')),
 ('Home and lifestyle', Decimal('911')),
 ('Fashion accessories', Decimal('902')),
 ('Health and beauty', Decimal('854'))]

The most selling product line is electronic accessories. The Electronic Accessories product line sold 971 units. The second most selling product line is Food and beverages.

#### *iv) What is the total revenue by month?*

Here, we group the sum of revenue by month.

In [60]:
%%sql
SELECT Month, ROUND(SUM(Total),2) AS TotalMonth
FROM Walmart
GROUP BY Month
ORDER BY TotalMonth DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('January', Decimal('116291.87')),
 ('March', Decimal('109455.51')),
 ('February', Decimal('97219.37'))]

The data shows the sales for three months: January, February and March. The above output shows that January had the highest sales $ 116291.87.

The total sales in February were $ 97219.37.

The sales in March were $109455.51.

#### *v) What month had the largest COGS?*

To determine the month with the highest COGS, we need to group the total COGS by month, and then sort it by total COGS in descending order.

In [18]:
%%sql
SELECT Month, SUM(COGS) AS TotalCOGS
FROM Walmart
GROUP BY Month
ORDER BY TotalCOGS DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('January', Decimal('110754.16')),
 ('March', Decimal('104243.34')),
 ('February', Decimal('92589.88'))]

Jnauary had the highest total cost of goods sold (COGS), $ 110754.16

#### *vi) What product line had the largest revenue?*

Here, we need to group the sum of total revenue by product line, and then sort it by total revenue in descending order.

In [59]:
%%sql
SELECT Product_line, ROUND(SUM(total),2) AS TotalRevenue
FROM Walmart
GROUP BY Product_line
ORDER BY TotalRevenue DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
6 rows affected.


[('Food and beverages', Decimal('56144.84')),
 ('Sports and travel', Decimal('55122.83')),
 ('Electronic accessories', Decimal('54337.53')),
 ('Fashion accessories', Decimal('54305.90')),
 ('Home and lifestyle', Decimal('53861.91')),
 ('Health and beauty', Decimal('49193.74'))]

The output shows that Food and beverages product line had the highest total revenue, $ 56144.84.

#### *vii) What is the city with the largest revenue?*

Here, we need to group the total revenue by city and the sort it by total revenue in descending order.

In [58]:
%%sql
SELECT City, ROUND(SUM(Total),2) AS TotalRevenue
FROM Walmart
GROUP BY City
ORDER BY TotalRevenue DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('Naypyitaw', Decimal('110568.71')),
 ('Yangon', Decimal('106200.37')),
 ('Mandalay', Decimal('106197.67'))]

The products were sold in three cities. Naypyitaw had the highest total revenue; $ 110568.71.

The products were sold in three cities. Naypyitaw had the highest total revenue; $ 110568.71.

#### *viii) What product line had the largest VAT?*

Here, we group the VAT by Product line and then sort it by VAT in descending order.

In [23]:
%%sql
SELECT Product_line, AVG(Tax_5_Percent) AS AVG_VAT
FROM Walmart
GROUP BY Product_line
ORDER BY AVG_VAT DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
6 rows affected.


[('Home and lifestyle', Decimal('16.03033125')),
 ('Sports and travel', Decimal('15.81262952')),
 ('Health and beauty', Decimal('15.41157237')),
 ('Food and beverages', Decimal('15.36531034')),
 ('Electronic accessories', Decimal('15.22059706')),
 ('Fashion accessories', Decimal('14.52806180'))]

The output shows that the Home and lifestyle product line had the largest VAT.

#### *ix) Fetch each product line and add a column to those product lines showing "Good," and "Bad." Good if it is greater than average sales.*

Here, we need to use a CASE statement and Sub queries to group the average quantity based on product lines, and then assign them to two classes as shown below.

In [25]:
%%sql
SELECT Product_line,  
CASE 
WHEN AVG(Quantity) >= (SELECT AVG(Quantity) FROM Walmart) THEN "Good" ELSE
"bad" END AS Performance
FROM Walmart
GROUP BY Product_line;

 * mysql+mysqldb://root:***@localhost/Walmart
6 rows affected.


[('Food and beverages', 'bad'),
 ('Health and beauty', 'Good'),
 ('Sports and travel', 'Good'),
 ('Fashion accessories', 'bad'),
 ('Home and lifestyle', 'Good'),
 ('Electronic accessories', 'Good')]

The output shows that most of the product lines performed good except Food and Beverages, and Fashion accessories.

#### *x) Which branch sold more products than average product sold?*

We need to summarize the sum of quantities grouped by branch and then filter them to only have branches with more than average quantity.

Here, We need to use subqueries to fetch the correct data. In my understanding, we need to find the branch with a total quantity greater than the average quantity of all branches.

The first step is to determine the total products sold in each branch.

In [28]:
%%sql
SELECT Branch, SUM(Quantity) AS BranchQuantity
FROM Walmart
GROUP BY Branch;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('A', Decimal('1859')), ('C', Decimal('1831')), ('B', Decimal('1820'))]

Now, we need to determine the average quantity of the branches using a subquery.

In [30]:
%%sql
SELECT AVG(BranchQuantity)
FROM (SELECT Branch, SUM(Quantity) AS BranchQuantity
FROM Walmart
GROUP BY Branch) B;

 * mysql+mysqldb://root:***@localhost/Walmart
1 rows affected.


[(Decimal('1836.6667'),)]

Now we know that the average quantity is 1836.6667.

It is time to write a more complex subquery to determine the branch(es) with that sold more products than average product sold.

In [36]:
%%sql
SELECT Branch, BranchQuantity
FROM (SELECT Branch, SUM(Quantity) AS BranchQuantity
FROM Walmart
GROUP BY Branch) B
WHERE BranchQuantity > (SELECT AVG(BranchQuantity)
FROM (SELECT branch, SUM(Quantity) AS BranchQuantity
FROM Walmart
GROUP BY Branch) B)
GROUP BY Branch;

 * mysql+mysqldb://root:***@localhost/Walmart
1 rows affected.


[('A', Decimal('1859'))]

Only branch A sold more products than the average products sold in each branch.

#### *xi) What is the most common product line by gender?*

Here we need to group the count of the InvoiceID by Gender, and then by Product line.

In [38]:
%%sql
SELECT Gender, Product_line, COUNT(InvoiceID) AS No_of_invoices
FROM Walmart
GROUP BY Gender, Product_line
ORDER BY No_of_invoices DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
12 rows affected.


[('Female', 'Fashion accessories', 96),
 ('Female', 'Food and beverages', 90),
 ('Male', 'Health and beauty', 88),
 ('Female', 'Sports and travel', 88),
 ('Male', 'Electronic accessories', 86),
 ('Male', 'Food and beverages', 84),
 ('Female', 'Electronic accessories', 84),
 ('Male', 'Fashion accessories', 82),
 ('Male', 'Home and lifestyle', 81),
 ('Female', 'Home and lifestyle', 79),
 ('Male', 'Sports and travel', 78),
 ('Female', 'Health and beauty', 64)]

Female fashion accessories were the most sold product line, followed by Female Food and beverages.

#### *xii) What is the average rating of each product line?*

In [40]:
%%sql
	SELECT Product_line, ROUND(AVG(Rating), 2) AS AverageRating
	FROM Walmart
	GROUP BY Product_line
	ORDER BY AverageRating DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
6 rows affected.


[('Food and beverages', Decimal('7.11')),
 ('Fashion accessories', Decimal('7.03')),
 ('Health and beauty', Decimal('7.00')),
 ('Sports and travel', Decimal('6.92')),
 ('Electronic accessories', Decimal('6.92')),
 ('Home and lifestyle', Decimal('6.84'))]

The Food and beverages product line has the highest average rating while the Home and lifestyle product line has the lowest.

## Sales EDA

This section of EDA is conducted to extract insights from the company's Sales.

#### *i) Number of sales made in each time of the day per weekday*

We first need to determine the earliest time a purchase was made.

In [47]:
%%sql
SELECT MIN(Time) AS Earliest_Time
FROM Walmart;

 * mysql+mysqldb://root:***@localhost/Walmart
1 rows affected.


[(datetime.timedelta(seconds=36000),)]

This means that the earliest purchase in a day was made at 10:00:00.

We need to determine the latest time.

In [48]:
%%sql
SELECT MAX(Time) AS Latest_Time
FROM Walmart;

 * mysql+mysqldb://root:***@localhost/Walmart
1 rows affected.


[(datetime.timedelta(seconds=75540),)]

The latest time a purchase was 20:59:00.

This means that no Purchases were made past 20:59:00 or before 10:00:00.

We can, therefore, use this information to classify the time into Morning, Afternoon, and Evening.

We shall use the CASE statement to classify our time variable.

The sum of the sales quantity then needs to be grouped based on the weekday and time of the day.

To get insights about the qunatnity of sales made at each time of the day per weekday, we need to sort the information based on WeekDay and Quantity of Sales. 

In [17]:
%%sql
SELECT DAYNAME(Date) as WeekDay, 
CASE 
WHEN Time < "12:00:00" THEN "Morning"
WHEN Time < "17:00:00" THEN "Afternoon"
ELSE "Evening" END AS TimeOfTheDay, SUM(Quantity) AS QuantityOfSales
FROM Walmart
GROUP BY  WeekDay, TimeOfTheDay
ORDER BY WeekDay,QuantityOfSales DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
21 rows affected.


[('Friday', 'Afternoon', Decimal('368')),
 ('Friday', 'Evening', Decimal('250')),
 ('Friday', 'Morning', Decimal('140')),
 ('Monday', 'Afternoon', Decimal('329')),
 ('Monday', 'Evening', Decimal('193')),
 ('Monday', 'Morning', Decimal('116')),
 ('Saturday', 'Evening', Decimal('380')),
 ('Saturday', 'Afternoon', Decimal('378')),
 ('Saturday', 'Morning', Decimal('161')),
 ('Sunday', 'Afternoon', Decimal('343')),
 ('Sunday', 'Evening', Decimal('291')),
 ('Sunday', 'Morning', Decimal('144')),
 ('Thursday', 'Afternoon', Decimal('330')),
 ('Thursday', 'Evening', Decimal('246')),
 ('Thursday', 'Morning', Decimal('179')),
 ('Tuesday', 'Afternoon', Decimal('356')),
 ('Tuesday', 'Evening', Decimal('333')),
 ('Tuesday', 'Morning', Decimal('173')),
 ('Wednesday', 'Afternoon', Decimal('427')),
 ('Wednesday', 'Evening', Decimal('248')),
 ('Wednesday', 'Morning', Decimal('125'))]

As observed in the afternoon, the highest sales quantity was made in the afternoon on each day of the week.

Very few sales were done during morning hours.

#### *ii) Which of the customer types brings the most revenue?*

Here, we need to select the customer type variable and the sum of total revenue, then group it by customer type and sort by revenue in descending order.

In [56]:
%%sql
SELECT Customer_type, ROUND(SUM(Total), 2) AS Revenue
FROM Walmart
GROUP BY Customer_type
ORDER BY Revenue DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
2 rows affected.


[('Member', Decimal('164223.44')), ('Normal', Decimal('158743.31'))]

The member customers brought the highest Revenue; $ 164223.44.

#### *Which city has the largest tax percentage/ VAT (Value Added Tax)?*

Here, we need to group the average tax percentage by city and then sort it in descending order.

In [65]:
%%sql
SELECT City, ROUND(AVG(Tax_5_Percent),3) AS AverageTax
FROM Walmart
GROUP BY City
ORDER BY AverageTax DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('Naypyitaw', Decimal('16.052')),
 ('Mandalay', Decimal('15.232')),
 ('Yangon', Decimal('14.874'))]

Naypyitaw city had the highest Tax percentage.

#### *iv) Which customer type pays the most in VAT?*

To get the consumer paying the most VAT, we calculated the average VAT grouped by consumer type.

In [71]:
%%sql
SELECT Customer_type, ROUND(AVG(Tax_5_Percent),3) AS AverageTax
FROM Walmart
GROUP BY Customer_type
ORDER BY AverageTax DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
2 rows affected.


[('Member', Decimal('15.609')), ('Normal', Decimal('15.149'))]

There is no large difference between the tax rate from the two customer types.

The member customers paid slightly higher tax rate than the normal customers.

## Customers EDA 

The purpoe of this section is to conduct exploratory data analysis to understand the company's customers.

#### *i)	How many unique customer types does the data have?*

Here, ewe select distinct customer types.

In [72]:
%%sql
SELECT DISTINCT Customer_type
FROM Walmart;

 * mysql+mysqldb://root:***@localhost/Walmart
2 rows affected.


[('Normal',), ('Member',)]

There are two unique customer types; Normal and Member.

#### *ii)	How many unique payment methods does the data have?*

Here we need to select distinct payment methods.

In [75]:
%%sql
SELECT DISTINCT Payment
FROM Walmart;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('Credit card',), ('Ewallet',), ('Cash',)]

There are three payment methods; Credit card, Ewallet and Cash;

#### *iii)	What is the most common customer type?*

Here we select the count of InvoiceID grouped by customer type and sorted in descending order.

In [83]:
%%sql
SELECT Customer_type, COUNT(InvoiceID) AS NumberOfCustomers
FROM Walmart
GROUP BY Customer_type
ORDER BY NumberOfCustomers DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
2 rows affected.


[('Member', 501), ('Normal', 499)]

The number of member customers is greater than normal customers by 2.

#### *iv) What is the gender of most of the customers?*

In [6]:
%%sql
SELECT Gender, COUNT(InvoiceID) AS NumberOfCustomers
FROM Walmart
GROUP BY Gender
ORDER BY  NumberOfCustomers DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
2 rows affected.


[('Female', 501), ('Male', 499)]

Majority of the customers were female.

#### *v) What is the gender distribution per branch?*

In [13]:
%%sql
SELECT Branch, Gender, COUNT(InvoiceID) AS NoOfCustomers
FROM Walmart
GROUP BY Branch, Gender
ORDER BY Branch, NoOfCustomers DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
6 rows affected.


[('A', 'Male', 179),
 ('A', 'Female', 161),
 ('B', 'Male', 170),
 ('B', 'Female', 162),
 ('C', 'Female', 178),
 ('C', 'Male', 150)]

The number of Male customers was higher than that of Female customers in branches A and B. The number of Female customers was higher in branch C.

#### *vi) Which time of the day do customers give the most Ratings?*

Here, we use the CASE statement to classify different times of the day, then group the data based on average ratings.

In [21]:
%%sql
SELECT 
CASE 
WHEN Time < "12:00:00" THEN "Morning"
WHEN Time < "17:00:00" THEN "Afternoon"
ELSE "Evening" END AS TimeOfTheDay, AVG(Rating) AS AverageRating
FROM Walmart
GROUP BY  TimeOfTheDay
ORDER BY AverageRating DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
3 rows affected.


[('Afternoon', Decimal('7.002203')),
 ('Morning', Decimal('6.960733')),
 ('Evening', Decimal('6.941408'))]

Customers give the most ratings in the afternoons.

#### *vii) At what time of the day do customers give the most ratings per branch?*

Here, we only need to update the above script to include the ratings per branch,

In [25]:
%%sql
SELECT Branch,
CASE 
WHEN Time < "12:00:00" THEN "Morning"
WHEN Time < "17:00:00" THEN "Afternoon"
ELSE "Evening" END AS TimeOfTheDay, AVG(Rating) AS AverageRating
FROM Walmart
GROUP BY  Branch, TimeOfTheDay
ORDER BY Branch, AverageRating DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
9 rows affected.


[('A', 'Afternoon', Decimal('7.093671')),
 ('A', 'Morning', Decimal('7.005479')),
 ('A', 'Evening', Decimal('6.944954')),
 ('B', 'Morning', Decimal('6.891525')),
 ('B', 'Afternoon', Decimal('6.816197')),
 ('B', 'Evening', Decimal('6.787023')),
 ('C', 'Evening', Decimal('7.113913')),
 ('C', 'Afternoon', Decimal('7.079870')),
 ('C', 'Morning', Decimal('6.974576'))]

In branch A, customers gave the most ratings in the Afternoons.

In branch B, most ratings were given in the Morning hours.

In branch C, most ratings were given in the Evinings.

#### *viii) Which day of the week has the best average ratings?*

Here, we write a script to extract the day of the week from the date column. We then group the average ratings based on the day of the week and sort it.

In [29]:
%%sql
SELECT DAYNAME(Date) AS DayOfWeek, AVG(Rating) AS AverageRating
FROM Walmart
GROUP BY DayOfWeek
ORDER BY AverageRating DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
7 rows affected.


[('Monday', Decimal('7.153600')),
 ('Friday', Decimal('7.076259')),
 ('Sunday', Decimal('7.011278')),
 ('Tuesday', Decimal('7.003165')),
 ('Saturday', Decimal('6.901829')),
 ('Thursday', Decimal('6.889855')),
 ('Wednesday', Decimal('6.805594'))]

Monday had the highest average rating.

#### *xi) Which day of the week has the best average ratings per branch?*

Here, we modify the above script to group by both branch and day of the week.

In [31]:
%%sql
SELECT Branch, DAYNAME(Date) AS DayOfWeek, AVG(Rating) AS AverageRating
FROM Walmart
GROUP BY Branch, DayOfWeek
ORDER BY Branch, AverageRating DESC;

 * mysql+mysqldb://root:***@localhost/Walmart
21 rows affected.


[('A', 'Friday', Decimal('7.312000')),
 ('A', 'Monday', Decimal('7.097917')),
 ('A', 'Sunday', Decimal('7.078846')),
 ('A', 'Tuesday', Decimal('7.058824')),
 ('A', 'Thursday', Decimal('6.958696')),
 ('A', 'Wednesday', Decimal('6.916279')),
 ('A', 'Saturday', Decimal('6.746000')),
 ('B', 'Monday', Decimal('7.335897')),
 ('B', 'Tuesday', Decimal('7.001887')),
 ('B', 'Sunday', Decimal('6.888571')),
 ('B', 'Thursday', Decimal('6.752273')),
 ('B', 'Saturday', Decimal('6.736667')),
 ('B', 'Friday', Decimal('6.694118')),
 ('B', 'Wednesday', Decimal('6.452000')),
 ('C', 'Friday', Decimal('7.278947')),
 ('C', 'Saturday', Decimal('7.229630')),
 ('C', 'Wednesday', Decimal('7.064000')),
 ('C', 'Monday', Decimal('7.036842')),
 ('C', 'Sunday', Decimal('7.028261')),
 ('C', 'Tuesday', Decimal('6.951852')),
 ('C', 'Thursday', Decimal('6.950000'))]

For branch A, Fridays have the highest average ratings.

For branch B, Mondays have the highest average rating.

For branch C, Fridays have the highest average ratings.

# Conclusion

This analysis has helped extract valuable insights from the dataset. Walmart management can use the results to understand their best-performing product lines and branches.

The management can use the information to understand the days of the week when most customers purchase products.

The firm can conduct an investigation to understand why customer ratings on Wednesdays and Thursdays are poor.